src/mame/sony/news_020_mmu.cpp

```diff
@@ -0,0 +1,298 @@
+{
+	(offset & 0x10000000) ? clear_kernel_entries() : clear_user_entries();
+}
+
+void news_020_mmu_device::clear_user_entries()
+{
+	LOGMASKED(LOG_ENTRY, "(%s) MMU clearing user entries\n", machine().describe_context());
+
+	for (int i = 0; i < MMU_ENTRY_COUNT; ++i)
+	{
+		m_mmu_user_ram[i] = 0x0;
+		m_mmu_user_tag_ram[i] = 0x0;
+	}
+}
+
+void news_020_mmu_device::clear_kernel_entries()
+{
+	LOGMASKED(LOG_ENTRY, "(%s) MMU clearing kernel entries\n", machine().describe_context());
+
+	for (int i = 0x0; i < MMU_ENTRY_COUNT; ++i)
+	{
+		m_mmu_system_ram[i] = 0x0;
+		m_mmu_system_tag_ram[i] = 0x0;
+	}
+}
+
+void news_020_mmu_device::device_start()
+{
+	m_bus_error.resolve();
+
+	save_item(NAME(m_enabled));
+	save_item(NAME(m_romdis));
+
+	m_mmu_user_ram = std::make_unique<u32[]>(MMU_ENTRY_COUNT);
+	save_pointer(NAME(m_mmu_user_ram), MMU_ENTRY_COUNT);
+	m_mmu_system_ram = std::make_unique<u32[]>(MMU_ENTRY_COUNT);
+	save_pointer(NAME(m_mmu_system_ram), MMU_ENTRY_COUNT);
+
+	m_mmu_user_tag_ram = std::make_unique<u32[]>(MMU_ENTRY_COUNT);
+	save_pointer(NAME(m_mmu_user_tag_ram), MMU_ENTRY_COUNT);
+	m_mmu_system_tag_ram = std::make_unique<u32[]>(MMU_ENTRY_COUNT);
+	save_pointer(NAME(m_mmu_system_tag_ram), MMU_ENTRY_COUNT);
+}
+
+device_memory_interface::space_config_vector news_020_mmu_device::memory_space_config() const
+{
+	return space_config_vector{
+		std::make_pair(AS_PROGRAM, &m_hyperbus_config)};
+}
+
+uint32_t news_020_mmu_device::mmu_entry_r(offs_t offset, uint32_t mem_mask)
+{
+	fatalerror("news_020_mmu: tried to read from mapping RAM!");
+	return 0;
+}
+
+void news_020_mmu_device::mmu_entry_w(offs_t offset, uint32_t data, uint32_t mem_mask)
+{
+	const bool system_map = offset & 0x4000000; // user or system mapping
+	const uint32_t tag = (offset & ~0x4000000);
+
+	// Actually update map
+	offset = offset % MMU_ENTRY_COUNT;
+	if (system_map)
+	{
+		COMBINE_DATA(&m_mmu_system_ram[offset]);
+		m_mmu_system_tag_ram[offset] = (m_mmu_system_tag_ram[offset] & ~mem_mask) | (tag & mem_mask);
+	}
+	else
+	{
+		COMBINE_DATA(&m_mmu_user_ram[offset]);
+		m_mmu_user_tag_ram[offset] = (m_mmu_user_tag_ram[offset] & ~mem_mask) | (tag & mem_mask);
+	}
+
+#if (VERBOSE & LOG_ENTRY) > 0
+	news_020_pte modified_entry(system_map ? m_mmu_system_ram[offset] : m_mmu_user_ram[offset]);
+	LOGMASKED(LOG_ENTRY, "(%s) MMU %s entry write (0x%08x, 0x%08x, 0x%08x) -> entry 0x%08x: tag = 0x%08x, pfnum = 0x%08x (valid = %d, modified = %d, fill_on_demand = %d, kernel write = %d, kernel read = %d, user write = %d, user read = %d)\n",
+			  machine().describe_context(),
+			  system_map ? "system" : "user",
+			  offset,
+			  data,
+			  mem_mask,
+			  offset % MMU_ENTRY_COUNT,
+			  tag,
+			  modified_entry.pfnum(),
+			  modified_entry.valid(),
+			  modified_entry.modified(),
+			  modified_entry.fill_on_demand(),
+			  modified_entry.kernel_writable(),
+			  modified_entry.kernel_readable(),
+			  modified_entry.user_writable(),
+			  modified_entry.user_readable());
+#endif
```

Comment:
Do you really need the `#if` here?  The compiler optimises away `LOGMASKED` to nothing if it’s disabled, and the only side effect of construction of the `news_020_pte` is a RAM access, which can be eliminated as the pointers aren’t volatile or atomic or anything like that.  Using the `#if` means the code is liable to rot because the compiler never sees it when it’s disabled.

Reply:
It wasn't needed, I wasn't thinking of the compiler optimizations. I removed it.